source/async_stream.c

```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/io/async_stream.h>
+
+#include <aws/common/byte_buf.h>
+#include <aws/io/future.h>
+#include <aws/io/stream.h>
+
+void aws_async_input_stream_init_base(
+    struct aws_async_input_stream *stream,
+    struct aws_allocator *alloc,
+    const struct aws_async_input_stream_vtable *vtable,
+    void *impl) {
+
+    AWS_ZERO_STRUCT(*stream);
+    stream->alloc = alloc;
+    stream->vtable = vtable;
+    stream->impl = impl;
+    aws_ref_count_init(&stream->ref_count, stream, (aws_simple_completion_callback *)vtable->destroy);
+}
+
+struct aws_async_input_stream *aws_async_input_stream_acquire(struct aws_async_input_stream *stream) {
+    if (stream != NULL) {
+        aws_ref_count_acquire(&stream->ref_count);
+    }
+    return stream;
+}
+
+struct aws_async_input_stream *aws_async_input_stream_release(struct aws_async_input_stream *stream) {
+    if (stream) {
+        aws_ref_count_release(&stream->ref_count);
+    }
+    return NULL;
+}
+
+struct aws_future_bool *aws_async_input_stream_read(struct aws_async_input_stream *stream, struct aws_byte_buf *dest) {
+    /* Deal with this edge case here, instead of relying on every implementation to do it right. */
+    if (dest->len == dest->capacity) {
+        struct aws_future_bool *future = aws_future_bool_new(stream->alloc);
+        aws_future_bool_set_error(future, AWS_ERROR_SHORT_BUFFER);
+        return future;
+    }
+
+    return stream->vtable->read(stream, dest);
+}
+
+/* Data to perform the aws_async_input_stream_read_to_fill() job */
+struct aws_async_input_stream_fill_job {
+    struct aws_allocator *alloc;
+    struct aws_async_input_stream *stream;
+    struct aws_byte_buf *dest;
+    /* Future for each read() step */
+    struct aws_future_bool *read_future;
+    /* Future to set when this job completes */
+    struct aws_future_bool *my_future;
+};
+
+static void s_async_stream_fill_job_complete(struct aws_async_input_stream_fill_job *job, bool eof, int error_code) {
+    if (error_code) {
+        aws_future_bool_set_error(job->my_future, error_code);
+    } else {
+        aws_future_bool_set_result(job->my_future, eof);
+    }
+    aws_future_bool_release(job->my_future);
+    aws_async_input_stream_release(job->stream);
+    aws_mem_release(job->alloc, job);
+}
+
+/* Call read() in a loop.
+ * It would be simpler to set a completion callback for each read() call,
+ * but this risks our call stack growing large if there are many small, synchronous, reads.
+ * So be complicated and loop until a read() ) call is actually async,
+ * and only then set the completion callback (which is this same function, where we resume looping). */
+static void s_async_stream_fill_job_loop(void *user_data) {
+    struct aws_async_input_stream_fill_job *job = user_data;
+
+    while (true) {
```

Comment:
If the underlying stream is sync and data is not available, we will just hang here. Should we introduce a time-off / sleep of some sort? Recently, Swift had a similar issue where CPU usage went to 100%.

Reply:
Thanks for pointing this out. aws_input_stream allows 0 byte reads, but the async stream does not. I fixed `s_async_wrapping_synchronous_stream_read()` so that it calls aws_input_stream_read() in a loop if it gets a 0 byte read.

Regarding whether or not we should sleep: I don't think we should. If anyone is hitting this loop and the CPU is spinning at 100%, they should migrate to using an ACTUAL async stream, instead of the old synchronous stream API. This wasn't an option in the past, but now it is!